<a href="https://colab.research.google.com/github/YuliiaMan/python_for_ds_tasks/blob/main/hw_6_4_%D0%90%D0%BD%D0%B0%D0%BB%D1%96%D0%B7_%D0%90%D0%92_%D1%82%D0%B5%D1%81%D1%82%D1%83.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Аналіз A/B тесту

Маємо проаналізувати дані A/B тесту в популярній [грі Cookie Cats](https://www.facebook.com/cookiecatsgame). Це класична гра-головоломка в стилі «з’єднай три», де гравець повинен з’єднати плитки одного кольору, щоб очистити дошку та виграти рівень. На дошці також зображені співаючі котики :)

Під час проходження гри гравці стикаються з воротами, які змушують їх чекати деякий час, перш ніж вони зможуть прогресувати або зробити покупку в додатку. У цьому блоці завдань ми проаналізуємо результати A/B тесту, коли перші ворота в Cookie Cats було переміщено з рівня 30 на рівень 40. Зокрема, ми проаналізуємо вплив на утримання (retention) гравців. Тобто хочемо зрозуміти чи переміщення воріт на 10 рівнів пізніше якимось чином вплинуло на те, що користувачі перестають грати в гру раніше чи пізніше з точки зору кількості їх днів з моменту встановлення гри.

Будемо працювати з даними з файлу `cookie_cats.csv`. Змінні в даних наступні:

- userid - унікальний номер, який ідентифікує кожного гравця.
- version - чи потрапив гравець в контрольну групу (gate_30 - ворота на 30 рівні) чи тестову групу (gate_40 - ворота на 40 рівні).
- sum_gamerounds - кількість ігрових раундів, зіграних гравцем протягом першого тижня після встановлення
- retention_1 - чи через 1 день після встановлення гравець повернувся і почав грати?
- retention_7 - чи через 7 днів після встановлення гравець повернувся і почав грати?

Коли гравець встановлював гру, його випадковим чином призначали до групи gate_30 або gate_40.

1. Зчитайте дані АВ тесту у змінну `df` та виведіть середнє значення показника показник `retention_7` (утримання на 7 день) по версіям гри. Сформулюйте гіпотезу: яка версія дає краще утримання через 7 днів після встановлення гри?

In [1]:
import pandas as pd
import numpy as np

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
data_path = '/content/drive/MyDrive/datastat/cookie_cats.csv'
df = pd.read_csv(data_path)

In [6]:
mean_retention_7 = df.groupby('version')['retention_7'].mean()
mean_retention_7

version
gate_30    0.190201
gate_40    0.182000
Name: retention_7, dtype: float64

Версія gate_30 дає краще утримання через 7 днів після встановлення гри

2. Перевірте з допомогою z-тесту аналогічно до прикладу в лекції, чи дає якась з версій гри кращий показник `retention_7` на рівні значущості 0.05. Обчисліть також довірчі інтервали для двох вибірок. Виведіть результат у форматі:
```
z statistic: ...
p-value: ...
Довірчий інтервал 95% для групи control: [..., ...]
Довірчий інтервал 95% для групи treatment: [..., ...]
```
де замість `...` - обчислені значення. В якості висновка дайте відповідь на два питання:  
    1. чи є статистична значущою різниця між поведінкою користувачів у різних версіях гри?   
    2. чи перетинаються довірчі інтервали утримання користувачів з різних версій гри? Про що це каже?  
    
Зверніть увагу, в такому і схожому завданнях ми використовуєм `proportion` Z-тест. Це тому що в нас залежна змінна має бінарне значення (повернеться аби ні користувач, чи клікне або ні користувач в інших ситуаціях - всього два можливих значення в змінної: 0/1, True/False ). Якщо б ми вимірювали скажімо чи є стат. значущою різниця між вагою чоловіків і жінок в певній вибірці, ми б використовувавли функцію `statsmodels.stats.ztest`, бо залежна змінна `вага` є неперервною (тип float, замість типу int чи bool і тільки двох можливих значень).

In [7]:
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

# Кількість гравців у кожній групі
n_control = df[df['version'] == 'gate_30']['retention_7'].count()
n_treatment = df[df['version'] == 'gate_40']['retention_7'].count()

# Кількість гравців, які повернулися через 7 днів
success_control = df[df['version'] == 'gate_30']['retention_7'].sum()
success_treatment = df[df['version'] == 'gate_40']['retention_7'].sum()

# Z-тест
successes = [success_control, success_treatment]
nobs = [n_control, n_treatment]

z_stat, pval = proportions_ztest(successes, nobs=nobs)
(conf_int_control, conf_int_treatment) = proportion_confint(successes, nobs=nobs, alpha=0.05)

print(f'z statistic: {z_stat:.2f}')
print(f'p-value: {pval:.3f}')
print(f'Довірчий інтервал 95% для групи control: {conf_int_control}')
print(f'Довірчий інтервал 95% для групи treatment: {conf_int_treatment}')

z statistic: 3.16
p-value: 0.002
Довірчий інтервал 95% для групи control: [0.18656312 0.1784543 ]
Довірчий інтервал 95% для групи treatment: [0.19383957 0.18554579]


Оскільки 𝑝-значення (0.002) < 0.05, це означає, що різниця між показниками утримання користувачів на 7-й день у двох версіях гри є статистично значущою.

Довірчі інтервали:
Для контрольної групи (gate_30): [0.178, 0.187]
Для тестової групи (gate_40): [0.186, 0.194]

Інтервали частково перетинаються у значенні 0.186

Перетин довірчих інтервалів свідчить, що хоча ми виявили статистично значущу різницю, вона не є дуже переконливою.

3. Є ще один тип тестів, який використовується для бінарної метрики як от "зробить юзер дію, чи ні" - тест [**Хі-квадрат**](https://www.bmj.com/about-bmj/resources-readers/publications/statistics-square-one/8-chi-squared-tests) (ще одне [пояснення](https://www.scribbr.com/statistics/chi-square-tests/) тесту з прикладами). В нього інші гіпотези Н0 і Н1 на відміну від z- та t-тестів. А також цей тест можна використовувати, якщо в нас більше за 2 досліджувані групи, тобто в нас не А/В тест, а А/B/C/D, наприклад.  

В **z- та t-тестах** (які відрізняються тим, що ми в першому не знаємо дисперсію генеральної сукупності, але якщо в нас великий набір даних, то ці два тести дають дуже схожі результати) **ми перевіряємо, чи є різниця у середніх показниках по групам користувачів**.  

А в **тесті Хі-квадрат ми перевіряємо чи є звʼязок між групою користувача і тим, чи він зробить цікаву нам дію**. Це ніби дослідження одного і того самого, але дещо різними способами. Для перевірки, можна виконувати кілька тестів (особливо, якщо один дає якийсь непереконливий результат типу р-значення 0.07 - наче і fail to regect H0 на рівні стат значущості 5%, але цікаво, що скажуть інші тести), тож, зробимо і ми тест хі-квадрат та порівняємо його результат з z-тестом.

Про різницю між тестами можна почитати ще [тут](https://stats.stackexchange.com/a/178860) - це просто пояснення користувача стековерфлоу, але там розумні люди сидять.

Для проведення хі-квадрат тесту скористаємось функцією з `scipy.stats` `chi2_contingency` для обчислення статистики хі-квадрат і р-значення для перевірки конкретної гіпотези. У цю функцію вам треба передати таблицю 2х2: кількість випадків для кожної версії гри і значення `retention_7`.

**Задача**: виконайте тест хі-квадрат на рівні значущості 5% аби визначити, чи є залежність між версією гри та тим, чи зайде гравець на 7ий день після встановлення гри.
Тут гіпотези наступні
- Н0: значення retention_7 не залежить від версії гри
- Н1: є залежність між версією гри і значенням retention_7

Виведіть p-значення та зробіть висновок.


In [8]:
from scipy.stats import chi2_contingency

contingency_table = [
    [success_control, n_control - success_control],
    [success_treatment, n_treatment - success_treatment]
]

chi2, pval_chi2, _, _ = chi2_contingency(contingency_table)
print(f'Chi-squared p-value: {pval_chi2:.3f}')

Chi-squared p-value: 0.002


p-value: 0.002. Оскільки p<0.05, то ми відхиляємо нульову гіпотезу Н0, яка стверджує, що утримання користувачів на 7-й день не залежить від версії гри.

Є статистично значуща залежність між версією гри (gate_30 або gate_40) та тим, чи повернеться гравець на 7-й день.
Це означає, що переміщення воріт з рівня 30 на рівень 40 впливає на утримання користувачів на 7-й день після встановлення гри.

Обидва тести (Z-тест і Хі-квадрат тест) дали схожий результат: є статистично значуща різниця між групами.
Це додає впевненості в тому, що зміна рівня воріт має вплив на поведінку користувачів.